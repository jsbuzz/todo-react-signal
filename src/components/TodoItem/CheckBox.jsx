import React from 'react';
import PropTypes from 'prop-types';

const CheckBox = ({ done, onChange }) => (
  <input
    type="checkbox"
    className="toggle"
    checked={done}
    onChange={onChange}
  />
);

CheckBox.propTypes = {
  done: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default CheckBox;
