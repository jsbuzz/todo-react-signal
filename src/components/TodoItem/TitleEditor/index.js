import Connect from '../../../../react-signal';
import TitleEditor from './TitleEditor';

export default Connect(TitleEditor);
