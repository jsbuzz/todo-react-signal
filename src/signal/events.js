import { Event } from '../../react-signal/event-hive/event';

export const AddTodoEvent = class extends Event {
  constructor(title) {
    super();
    this.title = title;
  }
};
export const IdEvent = class extends Event {
  constructor(id) {
    super();
    this.id = id;
  }
};

export const ErrorEvent = class extends Event {
  constructor(error) {
    super();
    this.error = error;
  }
};

export const AddTodo = AddTodoEvent.withAlias('Todo:add');
export const RemoveTodo = IdEvent.withAlias('Todo:remove');
export const ToggleTodo = IdEvent.withAlias('Todo:toggle');
