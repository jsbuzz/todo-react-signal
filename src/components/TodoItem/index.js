import Connect from '../../../react-signal';
import TodoItem from './TodoItem';

export default Connect(TodoItem);
