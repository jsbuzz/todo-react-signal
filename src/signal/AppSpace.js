import {
  NameSpace,
  InitState,
  set,
} from '../../react-signal/event-hive/namespace';
import { UpdateActive } from './events';

const AppSpace = NameSpace.schema('Ns.App', () => ({
  activeTodos: [
    InitState, set(0),
    
    // update number of active todos
    UpdateActive, activeTodos => ({ active }) => console.log(active) || active
  ]
}));

window.AppSpace = AppSpace;

export default AppSpace;
