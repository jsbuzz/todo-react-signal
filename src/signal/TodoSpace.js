import {
  NameSpace,
  InitState,
  set,
  modify
} from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, UpdateTodo, RestoreTodos } from './events';

const TodoSpace = NameSpace.schema((todoId = 0) => ({
  todos: [
    InitState, set(new Map()),
    
    // Add new todo
    AddTodo, modify(todos => ({ title }) => {
      const id = ++todoId;
      const todo = { id, title, done: false };

      todos.set(id, todo);
    }),
    
    // restore todos from cache
    RestoreTodos, modify(todos => ({ savedTodos }) => {
      savedTodos.forEach(todo => {
        todoId = todo.id;
        todos.set(todo.id, todo);
      });
    }),
    
    // remove a todo by id
    RemoveTodo, modify(todos => ({ id }) => todos.delete(id)),
    
    // update a todo by id
    UpdateTodo, modify(todos => ({ todo }) => todos.set(todo.id, todo))
  ]
}));

window.TodoSpace = TodoSpace;

export default TodoSpace;
