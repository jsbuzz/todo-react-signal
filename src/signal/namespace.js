import { NameSpace, InitState } from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, ToggleTodo } from './events';

const TodoSpace = NameSpace.get('Ns.Todos');

const findTodo = (todos, todoId) => {
  let i = 0;
  for (const todo of todos) {
    if (todo.id === todoId) {
      return i;
    }
    i += 1;
  }
  return -1;
};

const toggleTodo = (todo) => {
  todo.done = !todo.done;
}

let todoId = 0;
TodoSpace.defineState({
  todos: state => [
    InitState, () => (state.todos = []),
    AddTodo, ({ title }) => (state.todos.push({
      id: ++todoId,
      title,
      done: false,
    })),
    RemoveTodo, ({ id }) => (state.todos.splice(findTodo(state.todos, id), 1)),
    ToggleTodo, ({ id }) => (toggleTodo(state.todos[findTodo(state.todos, id)])),
  ],
});

export default TodoSpace;
