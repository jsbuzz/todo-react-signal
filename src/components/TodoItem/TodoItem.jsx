import React from 'react';
import PropTypes from 'prop-types';

import { RemoveTodo, UpdateTodo } from '../../signal/events';
import CheckBox from './CheckBox';
import TitleEditor from './TitleEditor';

const TodoItem = ({ todo }, ns) => {
  const { id, title, done, edited } = todo;
  return (
    <li
      className={(done ? 'todo completed' : 'todo') + (edited ? ' edited' : '')}
      onDoubleClick={() => ns().trigger(UpdateTodo.with({ id, title, done, edited: true }))}
    >
      <CheckBox
        onChange={() => ns().trigger(UpdateTodo.with({ id, title, done: !done }))}
        done={done}
      />
      <label className="todo-title">{title}</label>
      {edited && <TitleEditor todo={todo} />}
      <button
        type="button"
        className="destroy"
        onClick={() => ns().trigger(RemoveTodo.with(id))}
      >
      </button>
    </li>
  );
};


TodoItem.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    edited: PropTypes.bool,
    done: PropTypes.bool.isRequired,
  }).isRequired,
};

export default TodoItem;
