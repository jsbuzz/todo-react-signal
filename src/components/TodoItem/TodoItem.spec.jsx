import React from "react";
import { shallow } from "enzyme";
import {
  TestNameSpace,
  useTestNamespace
} from "../../../react-signal/test-connect";

import { UpdateTodo, RemoveTodo } from "../../signal/events";
import TodoItem from "./TodoItem";

describe("components/TodoItem", () => {
  const todo = {
    id: 191,
    title: "test me!",
    done: false,
    edited: false
  };
  let NS;
  let component;

  beforeEach(() => {
    NS = new TestNameSpace();

    useTestNamespace(NS);

    component = shallow(<TodoItem todo={todo} />);
  });

  describe("WHEN todo is marked as done", () => {
    beforeEach(() => {
      component.setProps({ todo: { ...todo, done: true } });
    });

    it("should match snapshot", () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe("WHEN todo is marked as NOT done", () => {
    it("should match snapshot", () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe("WHEN doubleClicked", () => {
    beforeEach(() => {
      component.find("li").simulate("doubleClick");
    });

    it("should trigger UpdateTodo with edited: true", () => {
      expect(NS.lastEvent).toBeInstanceOf(UpdateTodo);
      expect(NS.lastEvent.todo).toEqual({
        ...todo,
        edited: true
      });
    });
  });

  describe("WHEN checkbox is checked", () => {
    beforeEach(() => {
      component.find("CheckBox").simulate("change");
    });

    it("should trigger UpdateTodo with done: true", () => {
      expect(NS.lastEvent).toBeInstanceOf(UpdateTodo);
      expect(NS.lastEvent.todo).toEqual({
        ...todo,
        done: true
      });
    });
  });

  describe("WHEN checkbox is unchecked", () => {
    beforeEach(() => {
      component.setProps({ todo: { ...todo, done: true } });
      component.find("CheckBox").simulate("change");
    });

    it("should trigger UpdateTodo with done: false", () => {
      expect(NS.lastEvent).toBeInstanceOf(UpdateTodo);
      expect(NS.lastEvent.todo).toEqual({
        ...todo,
        done: false
      });
    });
  });

  describe("WHEN destroy button is clicked", () => {
    beforeEach(() => {
      component.find("button").simulate("click");
    });

    it("should trigger RemoveTodo with the right id", () => {
      expect(NS.lastEvent).toBeInstanceOf(RemoveTodo);
      expect(NS.lastEvent.id).toEqual(todo.id);
    });
  });
});
