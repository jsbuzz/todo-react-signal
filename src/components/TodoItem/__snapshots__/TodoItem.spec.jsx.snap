// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TodoItem WHEN todo is marked as NOT done should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoItem
    todo={
      Object {
        "done": false,
        "edited": false,
        "id": 191,
        "title": "test me!",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <CheckBox
          done={false}
          onChange={[Function]}
        />,
        <label
          className="todo-title"
        >
          test me!
        </label>,
        false,
        <button
          className="destroy"
          onClick={[Function]}
          type="button"
        />,
      ],
      "className": "todo",
      "onDoubleClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "done": false,
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "test me!",
          "className": "todo-title",
        },
        "ref": null,
        "rendered": "test me!",
        "type": "label",
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "destroy",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": null,
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <CheckBox
            done={false}
            onChange={[Function]}
          />,
          <label
            className="todo-title"
          >
            test me!
          </label>,
          false,
          <button
            className="destroy"
            onClick={[Function]}
            type="button"
          />,
        ],
        "className": "todo",
        "onDoubleClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "done": false,
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "test me!",
            "className": "todo-title",
          },
          "ref": null,
          "rendered": "test me!",
          "type": "label",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "destroy",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`components/TodoItem WHEN todo is marked as done should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoItem
    todo={
      Object {
        "done": true,
        "edited": false,
        "id": 191,
        "title": "test me!",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <CheckBox
          done={true}
          onChange={[Function]}
        />,
        <label
          className="todo-title"
        >
          test me!
        </label>,
        false,
        <button
          className="destroy"
          onClick={[Function]}
          type="button"
        />,
      ],
      "className": "todo completed",
      "onDoubleClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "done": true,
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "test me!",
          "className": "todo-title",
        },
        "ref": null,
        "rendered": "test me!",
        "type": "label",
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "destroy",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": null,
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <CheckBox
            done={true}
            onChange={[Function]}
          />,
          <label
            className="todo-title"
          >
            test me!
          </label>,
          false,
          <button
            className="destroy"
            onClick={[Function]}
            type="button"
          />,
        ],
        "className": "todo completed",
        "onDoubleClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "done": true,
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "test me!",
            "className": "todo-title",
          },
          "ref": null,
          "rendered": "test me!",
          "type": "label",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "destroy",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
